### DISCUSSION TITLE
How to use just the canvas and all its properties, detached from the UI?

### DISCUSSION THREAD

**Discussion by edthoms** [bitbof/klecks#33] [Q&A] [answered]
Hello, first I want to congratulate the team for making such an amazing tool. The drawback for it being so fully-featured is that the code is (understandably) laborious to fully comprehend at a glance, thus I am asking for a little bit of help. 

I would like to embed this on a webapp, but with a custom made UI around it, whilst keeping the awesome features native to the canvas itself, (like multitouch rotation, panning, and gestures in general) 

What directories/files should I focus on editing to remove the current UI but keep the "API endpoints" for actions like undo/redo, still reachable? 

Thanks!

**Comment by bitbof** [comment, marked as answer]
Hi,
This sounds interesting. The UI lives in:
- `app/script/klecks/brushes-ui`
- `app/script/klecks/filters` - unfortunately filters are still tied to their UI, but the non-ui part always sits in `apply(params)`
- `app/script/klecks/canvas-ui` - although it sounds like you want to keep some of its features
- `app/script/klecks/ui`

KlApp (`app/script/app/kl-app.ts`) is what creates and coordinates everything, and is unfortunately still a large script (almost 2k lines). In there, big things related to UI are:
- `toolspace` - The UI that sits next to the canvas. Everything that appends itself to it is UI too, so it won't be listed here.
- `toolspaceScroller` - allows toolspace to scroll when window not tall enough
- `overlayToolspace` - shows when pressing alt + ctrl
- `klCanvasWorkspace` - again, the component you might want to keep
- `statusOverlay` - status messages displayed at top of CanvasWorkspace (zoom & orientation indicator, etc.)
- `KL.KLImageDropper` - gets instantiated but not assigned to any variable. Shows when dragging image into window.
- `function importFinishedLoading()` - modals that show when importing an image
- `LANG` - I18n

Important to mention: drawing happens through brush-ui. The CanvasWorkspace emits the drawing gesture events, these go to brush-ui (each brush tab has its own UI object), which then interacts with the actual brushes. It's silly, but how it works at the moment, and should be easy to change.

Furthermore in main (`main-embed.ts`, `main-standalone.ts`) `saveReminder` is part controller, part UI.


-bitbof

**Reply by edthoms** [reply]
Thank you for your quick reply! Very helpful!
